### PR TITLE
Add harnesses for all public functions of `Layout`

### PR DESCRIPTION
**Pull request by tautschnig**
Exercise all public member functions of `Layout` and assert properties over the result. Some of those should, perhaps, become `ensures` clauses.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/core/src/alloc/layout.rs
Comment (celinval) [inline library/core/src/alloc/layout.rs:547]: Why not `proof_for_contract`?

File: library/core/src/alloc/layout.rs
Comment (celinval) [inline library/core/src/alloc/layout.rs:524]: Can you implement `kani::Arbitrary` and `kani::Invariant` for Layout instead?

File: library/core/src/alloc/layout.rs
Comment (celinval) [inline library/core/src/alloc/layout.rs:590]: For unsafe functions, it would be great if we could cover different safety scenarios.

**Review by feliperodri — COMMENTED** [review]
Some comments from https://github.com/model-checking/verify-rust-std/pull/42 also apply here. So I think maybe we should merge that previous PR first and implement some of the cleanups requested there in this one (e.g., using `kani::Arbitrary` and moving assertions to ensures clauses).

File: library/core/src/alloc/layout.rs
Comment (tautschnig) [reply library/core/src/alloc/layout.rs:590]: Hmm, yes, but how many would we consider "good enough?" And for this one specifically I believe we are lacking the means to express the different cases as `requires` clauses that the function's documentation calls out.

File: library/core/src/alloc/layout.rs
Comment (tautschnig) [reply library/core/src/alloc/layout.rs:524]: See https://github.com/model-checking/verify-rust-std/pull/42#discussion_r1703916363

File: library/core/src/alloc/layout.rs
Comment (tautschnig) [reply library/core/src/alloc/layout.rs:547]: Same problem as https://github.com/model-checking/verify-rust-std/pull/42#discussion_r1703939852

File: library/core/src/alloc/layout.rs
Comment (celinval) [reply library/core/src/alloc/layout.rs]: I don't have a number for you, but I would consider in this case adding a few cases with types that have different layouts

Comment (tautschnig) [reply library/core/src/alloc/layout.rs]: I have added another one, but need some help in producing objects for further two cases. TODOs added.

File: library/core/src/alloc/layout.rs
Comment (celinval) [inline library/core/src/alloc/layout.rs:350]:
```suggestion
    #[cfg_attr(not(kani), ensures(|result| result.is_err() || n == 0 || result.as_ref().unwrap().0.size() % n == 0))]
    #[cfg_attr(kani, ensures(|result| result.is_err() || n == 0 || result.as_ref().unwrap().0.size() >= self.size()))]
```

**Review by celinval — COMMENTED** [review]
I haven't finished yet

File: library/core/src/alloc/layout.rs
Comment (tautschnig) [reply library/core/src/alloc/layout.rs:180]: I don't agree this is redundant: this is a sanity check on what rustc produces.

File: library/core/src/alloc/layout.rs
Comment (tautschnig) [reply library/core/src/alloc/layout.rs:218]: How would that interact with contracts? Isn't https://model-checking.github.io/kani/crates/doc/kani/invariant/trait.Invariant.html specific to Kani, so would not be of any use for verification approaches (unlike the above `ensures` clause, which is meant to eventually become standard)?

File: library/core/src/alloc/layout.rs
Comment (tautschnig) [reply library/core/src/alloc/layout.rs:162]: Perhaps, but it's about the only thing we can add as contract to this function.

File: library/core/src/alloc/layout.rs
Comment (celinval) [reply library/core/src/alloc/layout.rs:221]: We should add an interface like we did with ensures

File: library/core/src/alloc/layout.rs
Comment (celinval) [reply library/core/src/alloc/layout.rs:162]: This function has no requirement, and it should always succeed. So I think we should remove the `requires`, which is the same as writing `#[requires(true)]`

File: library/core/src/alloc/layout.rs
Comment (celinval) [reply library/core/src/alloc/layout.rs:180]: I don't think contracts is meant to check if the compiler is working properly.

File: library/core/src/alloc/layout.rs
Comment (celinval) [reply library/core/src/alloc/layout.rs:221]: for now, you could use `#[cfg_attr(kani, ensures(|result| result.is_valid()))]`

**Comment by rahulku** [comment]
is this blocked on something?

File: library/core/src/alloc/layout.rs
Comment (tautschnig) [reply library/core/src/alloc/layout.rs]: I tried to use `Invariant` per the documentation, but got:
```
error[E0404]: expected trait, found derive macro `kani::Invariant`
   --> /verify-rust-std.git/library/core/src/alloc/layout.rs:583:10
    |
583 |     impl kani::Invariant for Layout {
    |          ^^^^^^^^^^^^^^^ not a trait

```

Comment (tautschnig) [reply library/core/src/alloc/layout.rs]: It seems that `Invariant` is not in `kani_core`.

Comment (celinval) [reply library/core/src/alloc/layout.rs]: @jaisnan

**Review by celinval — APPROVED** [review]
Sorry, I was hoping the other stuff would be in by now. So let's unblock this PR and we can go back and improve this once the Invariant PR merges.

**Review by carolynzech — APPROVED** [review]

**Pull request merged**
